[ORCA-302] Enable Automation

BWMac opened the pull request:
**Problem:**

Because previously both `nf-synapse-challenge` workflows were polling submission views for new submissions before evaluating them, we were concerned that once we have Airflow automating Tower runs of the workflows, the same submissions could be sent for evaluation more than once, causing repeated work in the best case, or conflicts and race conditions in the worst case.

**Solution:**

Modify our workflows to accept specific submission IDs as input instead of polling submission views. This way, as long as the correct submissions are given to a Tower run (whether manually or by Airflow) submissions should only be evaluated once. Users can now provide either a `manifest` file with submission IDs or a comma-separated string of `submissions` IDs. I opted for allowing both options because I am not sure what the maximum length of an input string for a Nextflow workflow is and we may be evaluating 100+ submissions at a time. The general idea is for manifests to be used when this is automated, but for submission strings to be used for testing (this way you don't need to stage a manifest in S3 for Tower tests)

**Notes:**
- Converted `GET_SUBMISSIONS` to `GET_SUBMISSION_IMAGES`. Now that we already have the IDs we can just use them to retrieve specific docker images in `MODEL_TO_DATA` runs.
- Added `-f` to the challenge annotation command because of this [bug](https://sagebionetworks.jira.com/browse/ORCA-302?focusedCommentId=198969).
- Updated many processes to take the `submissions_ch` channel as input and avoid some `map`-ing
- Updated a variable name in`RUN_DOCKER` to avoid confusion between containers.
- Added example CSV manifests for each workflow in `assets`
- Added a `subworkflow` `CREATE_SUBMISSION_CHANNEL` to create the workflow channel from either the manifest or submissions string. This is imported and used in both workflows (which have now been moved to `workflows` because I thought it made more sense that way.
- Added `maxForks 1` to `CREATE_FOLDERS` to prevent multiple folders with the same name from being created on Synapse at the same time in the event of >1 submission from the same submitter in one workflow run.
- Updated documentation to reflect changes (please look at the flow charts closely to ensure I connected things correctly).

**Testing on Tower:**
- `MODEL_TO_DATA_CHALLENGE` with [manifest](https://tower-dev.sagebionetworks.org/orgs/Sage-Bionetworks/workspaces/example-dev-project/watch/WT50ks5ocZ49f)
- `MODEL_TO_DATA_CHALLENGE` with [submissions](https://tower-dev.sagebionetworks.org/orgs/Sage-Bionetworks/workspaces/example-dev-project/watch/2bkTBN1AiAzjtH)
-  `DATA_TO_MODEL_CHALLENGE` with [manifest](https://tower-dev.sagebionetworks.org/orgs/Sage-Bionetworks/workspaces/example-dev-project/watch/2iSR6MxBIsIhrK)
-  `DATA_TO_MODEL_CHALLENGE` with [submissions](https://tower-dev.sagebionetworks.org/orgs/Sage-Bionetworks/workspaces/example-dev-project/watch/1UhLE9Uc0NNvaQ)

BWMac committed:
force annotation

BWMac committed:
adds send email to DAG

BWMac committed:
adds example manifests

BWMac committed:
updates workflows

BWMac committed:
updates get submissions

BWMac committed:
remove unessessary tuples

BWMac committed:
remove old submissions script

BWMac committed:
adds create_folders maxforks

BWMac committed:
updates manifest

BWMac committed:
make RUN_DOCKER dependent on CREATE_FOLDERS

BWMac committed:
updates data to model workflow

BWMac committed:
adds create submission channel workflow

BWMac committed:
update main.nf

BWMac committed:
update main workflows

BWMac committed:
remove channel wf

BWMac committed:
adds back submission wf

BWMac committed:
updates README

BWMac commented on workflows/DATA_TO_MODEL.nf:
```diff
@@ -29,18 +33,15 @@ include { ANNOTATE_SUBMISSION as ANNOTATE_SUBMISSION_AFTER_SCORE } from '../modu
-        .map { row -> tuple(row.submission_id, row.image_id) }
-    UPDATE_SUBMISSION_STATUS_BEFORE_RUN(image_ch.map { tuple(it[0], "EVALUATION_IN_PROGRESS") })
-    DOWNLOAD_SUBMISSION(image_ch.map {it[0]})
+    UPDATE_SUBMISSION_STATUS_BEFORE_EVALUATION(submission_ch, "EVALUATION_IN_PROGRESS")
```
One thing I realized is that this will be redundant during automated runs (since Airflow will do this). I kinda think we should leave it in for non-automated runs though.

jaymedina replied:
Yeah I don't see the harm in leaving this in for the manual runs. The whole procedure should be consistent between both types of runs

jaymedina commented on README.md:
```diff
@@ -21,60 +21,79 @@ This workflow expects a secret called `SYNAPSE_AUTH_TOKEN` (a Synapse Authentica
-3. One or more Docker images have already been submitted to your evaluation queue.
-4. Created a submission view that includes the `id` and `status` columns.
-5. Added the input data for evaluating submissions to a folder within your Synapse project.
+1. Created an evaluation queue within the Synapse project.
```
All of the numbering was changed to 1?

BWMac replied:
If you look at the rendering it numbers them correctly. This is a markdown trick to make it so you can add more steps without changing all of the numbers.

jaymedina replied:
Interesting! That's good to know

jaymedina commented on README.md:
```diff
@@ -21,60 +21,79 @@ This workflow expects a secret called `SYNAPSE_AUTH_TOKEN` (a Synapse Authentica
 The workflow takes the following inputs:
 
+1. `submissions` (required): A comma separated lis tof submission IDs to evaluate.
+1. `manifest` (required): A path to a submission manifest containing submissions IDs to evaluate.
```
I feel like this reads as if both `submissions` and `manifest` are necessary inputs, when you really just need one or the other. Can we make this "`submissions` or `manifest` (required):" or format this in another way

BWMac replied:
I changed "required" to "required  if `manifest/submission` is not provided"

jaymedina commented on README.md:
```diff
@@ -21,60 +21,79 @@ This workflow expects a secret called `SYNAPSE_AUTH_TOKEN` (a Synapse Authentica
+1. `scoring_script` (optional): The string name of the scoring script to use for the `SCORE` step of the workflow run. Defaults to `model_to_data_score.py`
+1. `validation_script` (optional): The string name of the validation script to use for the `VALIDATE` step of the workflow run. Defaults to `validate.py`
+
+***Note:*** You must provide one of `submissions` or `manifest`. If you provide both, `submissions` will take precedence. Generally, `submissions` should be used for testing and `manifest` for automation.
```
I see this note here now. If we decide not to reformat to `submissions or manifest`, can we at least move this to the top before the required fields are listed?

BWMac replied:
Moved to the top in both sections

jaymedina commented on README.md:
````diff
@@ -21,60 +21,79 @@ This workflow expects a secret called `SYNAPSE_AUTH_TOKEN` (a Synapse Authentica
+Run the workflow locally with default inputs and a `submissions` string input:
 ```
-nextflow run main.nf -entry MODEL_TO_DATA_CHALLENGE -profile local
+nextflow run main.nf -entry MODEL_TO_DATA_CHALLENGE -profile local --submissions 9741046,9741047
````
can we also add an example for the `manifest` input for completion? i.e.

```
nextflow run main.nf -entry MODEL_TO_DATA_CHALLENGE -profile local --manifest path/to/manifest.csv
```

BWMac replied:
added in latest commit

jaymedina commented on README.md:
```diff
@@ -21,60 +21,79 @@ This workflow expects a secret called `SYNAPSE_AUTH_TOKEN` (a Synapse Authentica
+    K-->M[UPDATE STATUS];
+    L-->N;
+    M-->N[SEND EMAIL];
+    N-->O[END];
```
very nice; I like how this flowchart illustrates that the main steps after the 4 first ones are `run docker` `validate` `score` `send email`.

jaymedina commented on README.md:
```diff
@@ -85,41 +104,44 @@ flowchart LR;
 
 The workflow requires the following inputs:
 
+1. `submissions` (required): A comma separated lis tof submission IDs to evaluate.
```
list of*

jaymedina replied:
Also I see that there's no note here about users only needing one or the other (`submissions` or `manifest`) and what happens when both are fed - can we also add that note in this section? I assume users will skip to whatever section is relevant to them, so duplicate documentation is good for those cases

BWMac replied:
Yeah good call, I just forgot to include it.

jaymedina commented on README.md:
````diff
@@ -85,41 +104,44 @@ flowchart LR;
+Run the workflow locally with default inputs and a `submissions` string input:
 ```
-nextflow run main.nf -entry DATA_TO_MODEL_CHALLENGE -profile local
+nextflow run main.nf -entry DATA_TO_MODEL_CHALLENGE -profile local --submissions 9741046,9741047
````
another example for `--manifest` here

BWMac replied:
added in latest commit

jaymedina commented on bin/get_submission_image.py:
```diff
@@ -0,0 +1,34 @@
+import synapseclient
+
+
+def get_submission_image(submission_id: str) -> pd.DataFrame:
```
nit: wrong Return type - should be string

jaymedina reviewed (changes requested):
Thanks for your work on this! Overall the workflows look cleaner and I also like the diagrams. Can you add a link to the tower runs pointing to this branch for both the MODEL_TO_DATA and DATA_TO_MODEL workflows, for record-keeping?

BWMac committed:
addresses review comments

BWMac commented:
@jaymedina Thanks for your review! I added Tower links in the description (I knew I was forgetting something!) and addressed your comments in my latest commits.

I'll also be holding off on merging this until after my work on Airflow is done, just in case any additional changes are needed

jaymedina reviewed (approved):
lgtm 🚀 thanks!

BryanFauble reviewed (approved):
🔥 LGTM!

BWMac committed:
pause emails for testing

BWMac committed:
bump memory for intensive processes

BWMac committed:
updates config
